src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -852,29 +852,61 @@ Shard::finalizeInternal(
         if (detShard)
             detShard->store(nodeObject);
 
-        std::lock_guard lock(mutex_);
-
-        // Remove the acquire SQLite database
-        if (acquireInfo_)
         {
-            acquireInfo_.reset();
-            remove_all(dir_ / AcquireShardDBName);
-        }
+            std::lock_guard lock(mutex_);
 
-        if (!initSQLite(lock))
-            return fail("failed to initialize SQLite databases");
+            // Remove the acquire SQLite database
+            if (acquireInfo_)
+            {
+                acquireInfo_.reset();
+                remove_all(dir_ / AcquireShardDBName);
+            }
 
-        setFileStats(lock);
-        lastAccess_ = std::chrono::steady_clock::now();
+            if (!initSQLite(lock))
+                return fail("failed to initialize SQLite databases");
+
+            setFileStats(lock);
+            lastAccess_ = std::chrono::steady_clock::now();
```

Comment:
This needs refactoring.

If `detShard` is valid then it is not necessary to call `initSQLite`, `setFilStats` here, that will be done in the call to `open`.
eg.
```
        if (detShard)
        {
            detShard->store(nodeObject);

            // Do not allow all other threads work with the shard
            busy_ = true;

            // Wait until all other threads leave the shard
            while (backendCount_ > 1)
                std::this_thread::yield();

            // Obtain the mutex
            std::lock_guard lock(mutex_);

            // Close original shard database
            backend_->close();

            // Close SQL databases for original shard
            lgrSQLiteDB_.reset();
            txSQLiteDB_.reset();

            // Remove the acquire SQLite database
            if (acquireInfo_)
            {
                acquireInfo_.reset();
                remove_all(dir_ / AcquireShardDBName);
            }

            // Close deterministic shard
            detShard->close();

            // Replace original shard by deterministic one.
            detShard->replace();

            // Re-open deterministic shard
            if (!open(lock))
                return false;

            // Allow all other threads work with the shard
            busy_ = false;
            lastAccess_ = std::chrono::steady_clock::now();
        }
        else
        {
            std::lock_guard lock(mutex_);

            // Remove the acquire SQLite database
            if (acquireInfo_)
            {
                acquireInfo_.reset();
                remove_all(dir_ / AcquireShardDBName);
            }

            if (!initSQLite(lock))
                return fail("failed to initialize SQLite databases");

            setFileStats(lock);
            lastAccess_ = std::chrono::steady_clock::now();
        }
```

Reply:
Done.